File: app/src/main/java/com/chatai/adr/util/DeviceHardwareUtil.java
package com.chatai.adr.util;

import android.app.ActivityManager;
import android.content.Context;
import java.util.Locale;

public class DeviceHardwareUtil {

    public static class MemoryStatus {
        public final long totalRamBytes;
        public final long availableRamBytes;
        public final boolean isLowMemory;
        public final int thresholdPercent;

        public MemoryStatus(long totalRamBytes, long availableRamBytes, boolean isLowMemory) {
            this.totalRamBytes = totalRamBytes;
            this.availableRamBytes = availableRamBytes;
            this.isLowMemory = isLowMemory;
            this.thresholdPercent = totalRamBytes > 0 ? (int) ((totalRamBytes - availableRamBytes) * 100 / totalRamBytes) : 0;
        }

        public double getTotalRamGB() {
            return (double) totalRamBytes / (1024.0 * 1024.0 * 1024.0);
        }

        public double getAvailableRamGB() {
            return (double) availableRamBytes / (1024.0 * 1024.0 * 1024.0);
        }

        public String getFormattedSummary() {
            return String.format(Locale.getDefault(), "RAM Khả dụng: %.1f GB / %.1f GB (Đã dùng: %d%%)",
                    getAvailableRamGB(), getTotalRamGB(), thresholdPercent);
        }
    }

    public static MemoryStatus getMemoryStatus(Context context) {
        if (context == null) {
            return new MemoryStatus(4L * 1024 * 1024 * 1024, 2L * 1024 * 1024 * 1024, false);
        }
        ActivityManager activityManager = (ActivityManager) context.getSystemService(Context.ACTIVITY_SERVICE);
        ActivityManager.MemoryInfo memoryInfo = new ActivityManager.MemoryInfo();
        if (activityManager != null) {
            activityManager.getMemoryInfo(memoryInfo);
            return new MemoryStatus(memoryInfo.totalMem, memoryInfo.availMem, memoryInfo.lowMemory);
        }
        return new MemoryStatus(4L * 1024 * 1024 * 1024, 2L * 1024 * 1024 * 1024, false);
    }

    /**
     * Checks whether the device has enough free RAM to safely load a local model
     * with a 500 MB safety buffer to prevent Android Out-Of-Memory (OOM) kills.
     */
    public static boolean canSafelyLoadModel(Context context, long requiredRamBytes) {
        MemoryStatus status = getMemoryStatus(context);
        long safetyBufferBytes = 500L * 1024 * 1024; // 500 MB buffer
        return status.availableRamBytes >= (requiredRamBytes + safetyBufferBytes);
    }
}

File: app/src/main/java/com/chatai/adr/util/DeviceHardwareUtil.kt
package com.chatai.adr.util

import android.app.ActivityManager
import android.content.Context
import java.util.Locale

object DeviceHardwareUtil {

    fun getTotalRamBytes(context: Context): Long {
        return try {
            val actManager = context.getSystemService(Context.ACTIVITY_SERVICE) as? ActivityManager
            val memInfo = ActivityManager.MemoryInfo()
            actManager?.getMemoryInfo(memInfo)
            memInfo.totalMem
        } catch (e: Exception) {
            4L * 1024 * 1024 * 1024 // Fallback 4GB
        }
    }

    fun getAvailableRamBytes(context: Context): Long {
        return try {
            val actManager = context.getSystemService(Context.ACTIVITY_SERVICE) as? ActivityManager
            val memInfo = ActivityManager.MemoryInfo()
            actManager?.getMemoryInfo(memInfo)
            memInfo.availMem
        } catch (e: Exception) {
            1500L * 1024 * 1024
        }
    }

    fun formatBytes(bytes: Long): String {
        val mb = bytes.toDouble() / (1024.0 * 1024.0)
        return if (mb >= 1024.0) {
            String.format(Locale.US, "%.1f GB", mb / 1024.0)
        } else {
            String.format(Locale.US, "%.0f MB", mb)
        }
    }

    fun getHardwareSummary(context: Context): String {
        val total = formatBytes(getTotalRamBytes(context))
        val avail = formatBytes(getAvailableRamBytes(context))
        return "RAM: $avail khả dụng / $total tổng"
    }

    fun canSafelyRunModel(context: Context, requiredVramBytes: Long): Boolean {
        val available = getAvailableRamBytes(context)
        val safetyMargin = 300L * 1024 * 1024 // 300MB buffer cho Android OS
        return available > (requiredVramBytes + safetyMargin)
    }
}
